onnxruntime/core/framework/execution_provider.cc

```diff
@@ -200,21 +200,36 @@ int IExecutionProvider::ModelMetadefIdGenerator::GenerateId(const onnxruntime::G
   } else {
     uint32_t hash[4] = {0, 0, 0, 0};
 
+    auto hash_str = [&hash](const std::string& str) {
+      MurmurHash3::x86_128(str.data(), gsl::narrow_cast<int32_t>(str.size()), hash[0], &hash);
+    };
+
     // prefer path the model was loaded from
     // this may not be available if the model was loaded from a stream or in-memory bytes
-    const auto& model_path_str = main_graph.ModelPath().ToPathString();
-    if (!model_path_str.empty()) {
-      MurmurHash3::x86_128(model_path_str.data(), gsl::narrow_cast<int32_t>(model_path_str.size()), hash[0], &hash);
-    } else {
-      auto hash_str = [&hash](const std::string& str) {
-        MurmurHash3::x86_128(str.data(), gsl::narrow_cast<int32_t>(str.size()), hash[0], &hash);
-      };
+    const auto& model_path = main_graph.ModelPath();
+    if (!model_path.IsEmpty()) {
+      std::string path_str;
+      if (hash_model_name) {
+        PathString leaf = model_path.GetComponents().back();
+        path_str = ToMBString(leaf.c_str());
+      } else {
+        path_str = ToMBString(model_path.ToPathString().c_str());
+      }
 
+      // ensure enough characters are hashed in case model names are short or very similar
+      int32_t path_length = gsl::narrow_cast<int32_t>(path_str.size());
+      int32_t string_length = 0, hash_string_length = 500;
```

Comment:
Should hash_string_length be a constexpr?

Reply:
Thanks. updated